EKF/mag_control.cpp

```diff
@@ -0,0 +1,355 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2019 Estimation and Control Library (ECL). All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name ECL nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file mag_control.cpp
+ * Control functions for ekf magnetic field fusion
+ */
+
+#include "ekf.h"
+#include <mathlib/mathlib.h>
+
+void Ekf::controlMagFusion()
+{
+	// If we are on ground, store the local position and time to use as a reference
+	// Also reset the flight alignment flag so that the mag fields will be re-initialised next time we achieve flight altitude
+	if (!_control_status.flags.in_air) {
+		_last_on_ground_posD = _state.pos(2);
+		_control_status.flags.mag_aligned_in_flight = false;
+		_num_bad_flight_yaw_events = 0;
+	}
+
+	if ((_params.mag_fusion_type >= MAG_FUSE_TYPE_NONE)
+	    || _control_status.flags.mag_fault
+	    || !_control_status.flags.tilt_align) {
```

Comment:
We already denied 3D fusion before tilt alignment: https://github.com/PX4/ecl/blob/7c1e38d3adc79db7586f86dba64f7b15635792b2/EKF/control.cpp#L1412
And it doesn't make sense to me to fuse magnetic heading if the tilt isn't aligned. Does that help to align tilt? I don't think so.
What do you think @priseborough  ?

Reply:
We shouldn't be using mag data before tilt is know because regardless of the method used, it requires comparison with earth frame field data. 

The _control_status.flags.yaw_align flag should be true before fusing magnetic data and the _control_status.flags.tilt_align flag should be true before aligning the yaw using magnetic data.

Reply:
The tilt_align flag is set true on initial filter alignment and ever goes false unless the EKF is restarted. It would  make more sense to use _control_status.flags.yaw_align as a requirement for fusing mag data.